design/upgrade.md

```diff
@@ -0,0 +1,165 @@
+But, the upgrade controller should be able to slowly increase its component update batch size as it proceeds through each component type.
+For example, in true canary deployment fashion, a single monitor could be upgraded to the new version and mon health/quorum will be verified.
+Then two monitors could be updated at once and verification occurs again, followed by four monitors, etc. up to a reasonable upper bound.
+We do not want too many pods going down at one time, which could potentially impact cluster health and functionality, so a sane upper bound will be important.
+
+### **Troubleshooting**
+If an upgrade does not succeed, especially if the rollback effort also fails, we want to have some artifacts that are accessible by the storage administrator to troubleshoot the issue or to reach out to the Rook community for help.
+Because the upgrade process involves terminating pods and starting new ones, we need some strategies for investigating what happened to pods that may no longer be alive.
+Listed below are a few techniques for accessing debugging artifacts from pods that are no longer running:
+* `kubectl logs --previous ${POD_NAME} ${CONTAINER_NAME}` allows you to retrieve logs from a previous instance of a pod (e.g. a pod that crashed but is not yet terminated)
+* `kubectl get pods --show-all=true` will list all pods, including older versioned pods that were terminated in order to replace them with pods running the newer version.
+* The Rook operator logs (which host the upgrade controller output) should be thorough and verbose about the following:
+  * The sequence of actions it took during the upgrade
+  * The replication controllers (e.g. daemon set, replica set, deployment) that it modified and the pod names that it terminated
+  * All health check status and output it encountered
+
+## **Next Steps**
+A theme of the `0.6` milestone is that "Rook is upgradeable", which we have demonstrated with the manual process outlined in the [Rook Upgrade User Guide](../Documentation/upgrade.md).
+Fully automated upgrade support has been described within this design proposal, but will likely need to be implemented in an iterative process, with lessons learned along the way from pre-production field experience.
+
+The next step will be to implement the happy path where the upgrade controller automatically updates all Rook components in the [described sequence](#general-sequence) and stops immediately if any health checks fail and the cluster does not return to a healthy functional state.
+
+Handling failure cases with rollback as well as handling migrations and breaking changes will likely be implemented in future milestones, along with reliability and stability improvements from field and testing experience.
+
+## **Open Questions**
+1. In the process of attempting to upgrade itself, if the operator pod (and therefore upgrade controller) do not come back up healthy, how can that be rectified?  Who can roll it back?
```

Comment:
this is another reason that upgrading the operator first should be done before updating clusters is done.

Reply:
yep, having the user update the operator first makes it easier/possible for them to make sure things are okay before any clusters get updated via automation.